proxy/proxy/client.go

```diff
@@ -292,33 +241,129 @@ func isExpired(cfg *tls.Config) bool {
+		if timeToRefresh <= 0 {
+			// If a new certificate expires before our buffer has expired,
+			// we should wait a bit and schedule a new refresh to much closer to the expiration's date
+			// TODO: This situation probably only occurs for Issue #622 - when the oauth2 token isn't refreshed before the cert is
+			timeToRefresh = certExpiration.Sub(now) - (5 * time.Second)
+			logging.Errorf("new ephemeral certificate expires sooner than expected (adjusting refresh time to compensate): current time: %v, certificate expires: %v", now, certExpiration)
+		}
+		logging.Infof("Scheduling refresh of ephemeral certifcate in %s", timeToRefresh)
+		go c.refreshCertAfter(instance, timeToRefresh)
+	}()
+	return done
+}
+
+// isValid returns true if the cacheEntry is still useable
+func isValid(c cacheEntry) bool {
+	// the entry is only valid there wasn't an error retrieving it and it has a cfg
+	return c.err == nil && c.cfg != nil
+}
+
+func needsRefresh(e cacheEntry, refreshCfgBuffer time.Duration) bool {
+	if e.done == nil { // no refresh started
+		return true
+	}
+	if !isValid(e) || time.Since(e.lastRefreshed) >= refreshCfgBuffer {
+		// if the entry is invalid or close enough to expiring check
+		// use the entry's done channel to determine if a refresh has started yet
+		select {
+		case <-e.done: // last refresh completed, so it's time for a new one
+			return true
+		default: // new refresh already started, so we can wait on that
+			return false
+		}
+	}
+	return false
+}
+
+func (c *Client) cachedCfg(instance string) (string, *tls.Config, string, error) {
 	c.cacheL.RLock()
-	ret, ok := c.cfgCache[instance]
+
+	throttle := c.RefreshCfgThrottle
+	if throttle == 0 {
+		throttle = DefaultRefreshCfgThrottle
+	}
+	refreshCfgBuffer := c.RefreshCfgBuffer
+	if refreshCfgBuffer == 0 {
+		refreshCfgBuffer = defaultRefreshCfgBuffer
+	}
+
+	e := c.cfgCache[instance]
 	c.cacheL.RUnlock()
+	if needsRefresh(e, refreshCfgBuffer) {
+		// Reenter the critical section with intent to make changes
+		c.cacheL.Lock()
+		if c.cfgCache == nil {
+			c.cfgCache = make(map[string]cacheEntry)
+		}
+		// the state may have changed between critical sections, so double check
+		e = c.cfgCache[instance]
+		if needsRefresh(e, refreshCfgBuffer) {
+			if time.Since(e.lastRefreshed) >= throttle {
+				// start a new refresh and update the cachedEntry to reflect that
+				e.done = c.startRefresh(instance, refreshCfgBuffer)
+				e.lastRefreshed = time.Now()
+				c.cfgCache[instance] = e
+			} else {
+				// TODO: Should this be a returned error?
+				logging.Errorf("Throttling refreshCfg(%s): it was only called %v ago", instance, time.Since(e.lastRefreshed))
+			}
+		}
+		c.cacheL.Unlock()
+	}
 
-	// Don't waste time returning an expired/invalid cert.
-	if !ok || ret.err != nil || isExpired(ret.cfg) {
-		return "", nil, ""
+	if !isValid(e) {
+		// if the previous result was invalid, wait for the next result to complete
+		// TODO: Ask for context?
```

Comment:
What do you mean with this TODO?

Reply:
Update cachedCfg to use a context arg (which I've just done).